### PR TITLE
citogo - a Go test wrapper that solves a bunch of CI problems

### PR DESCRIPTION
**Pull request by maxtaco**
- run one test in a process, so they don't leak threads and memory
- output failed logs to files, so the logs don't become insane                                
- in CI, output logs to S3, for the above reason                                   
- allow a certain number of flakes (3)                                            
- after 3 failures, don't keep running, just bail early.

### CODE REVIEW

**Comment by jzila** [comment]
I dont have time to look at this today, but please dont merge until I've taken a look tomorrow.

**Comment by maxtaco** [comment]
No rush! I think it's failing CI becuase "recursively fail to remove directory" is failing and it's the last thing in the run.

**Comment by maxtaco** [comment]
John, any idea why build 29 failed?  It's all green (except for workspace delete) but still the final build shows red

**Comment by patrickxb** [comment]
i think some stats -> stathat would be helpful to see how CI is doing.

**Comment by maxtaco** [comment]
this framework makes that totally feasible, since now we can notate individual test runs. We can add that in a subsequent PR

**Comment by maxtaco** [comment]
OK, this thing is ready to go and pretty well-tested at this point. Changes to `Jenksinsfile` are minimal

File: Jenkinsfile [hunk 1]
```diff
@@ -771,10 +773,16 @@ def testGo(prefix, packagesToTest) {
...
+              withCredentials([
+                [$class: 'StringBinding', credentialsId: 'citogo-flake-webhook', variable : 'CITOGO_FLAKE_WEBHOOK'],
+                [$class: 'StringBinding', credentialsId: 'citogo-aws-secret-access-key', variable : 'CITOGO_AWS_SECRET_ACCESS_KEY'],
+                [$class: 'StringBinding', credentialsId: 'citogo-aws-access-key-id', variable : 'CITOGO_AWS_ACCESS_KEY_ID']
+              ]) {
                 println "Running tests for ${testSpec.dirPath}"
-                sh "./${testBinary} -test.timeout ${testSpec.timeout}"
+                def t = getOverallTimeout(testSpec)
+                timeout(activity: true, time: t.time, unit: t.unit) {
+                  sh "citogo --flakes 3 --fails 3 --build ${env.BRANCH_NAME}_${env.BUILD_ID} --prefix ${testSpec.dirPath} --s3bucket ci-fail-logs --build-url ${env.BUILD_URL} --get-log-cmd /keybase/team/keybase/bin/ciget --no-compile --test-binary ./${testBinary}"
```
Comment (jzila) [inline Jenkinsfile]: What is this `--get-log-cmd` and how does it have access to the keybase team?

Comment (maxtaco) [reply Jenkinsfile]: it's for displaying how to get the log, but it's not actually getting the log

File: go/citogo/main.go [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+
+func (r *runner) testerName() string {
+	if r.opts.TestBinary != "" {
+		return r.opts.TestBinary
+	}
+	return fmt.Sprintf(".%c%s.test", os.PathSeparator, r.opts.DirBasename)
+}
+
+func (r *runner) listTests() error {
+	cmd := exec.Command(r.testerName(), "-test.list", ".")
```
Comment (jzila) [inline go/citogo/main.go]: TIL `-test.list` 🤯

**Pull request merged**